Add admin trees module

matiasmolleja opened the pull request:
Enables creation of admin menus from the admin.

matiasmolleja commented:
I made some improvements after the comments on the meeting.

![improvements01](https://user-images.githubusercontent.com/2589629/46821371-bd4d3480-cd88-11e8-98c0-c8160092f448.gif)

pgy866 commented:
Is the support system menu hidden in non-administrator rights?

sebastienros commented:
@pgy866 there is a permission that you need to be able to change it. But the change affects all users. We talked about adding a way for users to add their own links that would not be shared.

@matiasmolleja Do the links support custom permissions? I mean that when a link is added in the navigation provider, they should also check for permissions before being visible.

matiasmolleja commented:
Added custom icons feature.
I will look at the permissions used and fix what's needed.

![custom-icons](https://user-images.githubusercontent.com/2589629/46921420-47043880-cffb-11e8-9dce-02d409c5143f.gif)

sebastienros commented:
Why do we have css classes and an icon picker? Why not just the icon picker?

matiasmolleja commented:
> Why do we have css classes and an icon picker? Why not just the icon picker?
You can add a css class to style in different way a concrete admin link. I tried with a different background color for example. 
I just wanted to add an UI to the property that you set for the MenuItem class.

sebastienros commented:
What's an actual usage of it? Sounds very linked to what the admin theme can do, and the admin theme doesn't have anything right now to customize them.

matiasmolleja commented:
Right now it is not used anywhere. 
It could be used for things like changing colors, font-sizes, maybe changing all descendants menuItems style by setting a class in the UI...
But yes it is limited; it's only css.
Remove it?

matiasmolleja commented:
The default icon was added because the layout was broken when a first level menuItem was added without the icon :(

Plus, I think that even if aligned, the menuitem looks a little bit weird being the only one without an icon.
![sin titulo-1](https://user-images.githubusercontent.com/2589629/47038756-8ad57a00-d182-11e8-996e-961c509a7a3b.jpg)

sebastienros commented:
Yeah, I was expecting it to be ugly. Then just set a default one in the picker.

matiasmolleja commented:
But then, my default is going to be set, although not used, on any linkMenuitem, no matter what level it is on. I would prefer to leave it empty. And let TheAdmin take care of how it looks when it needs to show a link on the first level that has no icon set. Take care of looking good.

And the user still can remove the icon with the "trash" button. Or do you mean set a default and remove the button?

sebastienros commented:
I don't understand. What I meant is that instead of forcing a value if not is defined, put the circle as the default icon picker selection, but let the user click on trash to have nothing. The default is just to initialize the drop down on new links.

matiasmolleja commented:
OK. I understand. I will remove classes and set a default icon. If someone removes the default icon it will look ugly but I will be their fault :)

matiasmolleja commented:
I think that's it

sebastienros commented:
Can I have multiple root nodes in a single Admin Tree Node ?

matiasmolleja commented:
Yes, you can.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.AdminTrees/Controllers/NodeController.cs:
```diff
@@ -0,0 +1,349 @@
+                return Unauthorized();
+            }
+
+            var tree = await _session.GetAsync<AdminTree>(id);
```
This should go in a service class, and be cached and invalidated when it changes. Look at the site settings for an example on how to cache it in memory.

matiasmolleja replied:
Done

sebastienros commented on src/OrchardCore.Modules/OrchardCore.AdminTrees/Indexes/AdminTreeIndex.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace OrchardCore.AdminTrees.Indexes
+{
+    public class AdminTreeIndex : MapIndex
```
I don't think you need an index at all. You only have a single document.

matiasmolleja replied:
Done. I removed the index and the migration file too.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.AdminTrees/OrchardCore.AdminTrees.csproj:
```diff
@@ -0,0 +1,31 @@
+  <ItemGroup>
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Admin.Abstractions\OrchardCore.Admin.Abstractions.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.Module.Targets\OrchardCore.Module.Targets.csproj" />
+    <ProjectReference Include="..\..\OrchardCore\OrchardCore.ContentManagement\OrchardCore.ContentManagement.csproj" />
```
Modules don't reference modules. Just abstractions.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.AdminTrees/Services/AdminTreeNavigationProvidersCoordinator.cs:
```diff
@@ -0,0 +1,66 @@
+                return;
+            }
+
+            var trees = (await _session.Query<AdminTree>().ListAsync())
```
So apparently you store as many admin trees. But I think you just need a single document in the database. And have it cached in a custom service.

matiasmolleja replied:
Done.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.AdminTrees/Views/Items/LinkAdminNode.Fields.TreeSummary.cshtml:
```diff
@@ -0,0 +1,12 @@
+
+<div class="@(enabled ? "" : "disabled") concrete-link">
+    <h5>@T["Link"]</h5>
+    <span class="text-muted">@T["Link Text:"] @linkText | @T["Link Url:"]: @linkUrl</span>
```
`@T["Link Text: {0}", linkText] | @T["Link Url: {0}", linkUrl]`

Because in a different language the order might matter.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.AdminTrees/Views/Node/List.cshtml:
```diff
@@ -0,0 +1,135 @@
+            error: function (error) {
+                console.error(error.responseText);
+
+                alert("@T["There was an error when moving the tree node"]");
```
Don't use localization in JS, create a hidden div with a data attribute containing the text, and reference it using js. Otherwise a translation might have a `"` and break the JavaScript.

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Lists/AdminNodes/ListsAdminNodeNavigationBuilder.cs:
```diff
@@ -0,0 +1,103 @@
+            {
+                var cim = await _contentManager.PopulateAspectAsync<ContentItemMetadata>(ci);
+
+                if ((cim.AdminRouteValues.Any()) && (ci.DisplayText != null))
```
Remove extra parens

sebastienros commented on src/OrchardCore.Modules/OrchardCore.Lists/OrchardCore.Lists.csproj:
```diff
@@ -13,6 +13,7 @@
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.Media.Abstractions\OrchardCore.Media.Abstractions.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.ResourceManagement.Abstractions\OrchardCore.ResourceManagement.Abstractions.csproj" />
     <ProjectReference Include="..\..\OrchardCore\OrchardCore.MetaWeblog.Abstractions\OrchardCore.MetaWeblog.Abstractions.csproj" />
+    <ProjectReference Include="..\OrchardCore.AdminTrees\OrchardCore.AdminTrees.csproj" />
```
Modules don't reference modules. Create an abstraction project if you need to expose something. And add a feature dependency.

matiasmolleja replied:
Done. But not sure if I should add admintree as a dependency in the manifest file. I just added a project reference pointing to the abstraction. I did it for the List and Contents modules., of course.

sebastienros commented on src/OrchardCore/OrchardCore.Application.Cms.Targets/OrchardCore.Application.Cms.Targets.csproj:
```diff
@@ -33,6 +33,7 @@
     -->
 
     <ProjectReference Include="..\..\OrchardCore.Modules\OrchardCore.Admin\OrchardCore.Admin.csproj" PrivateAssets="none" />
+    <ProjectReference Include="..\..\OrchardCore.Modules\OrchardCore.AdminTrees\OrchardCore.AdminTrees.csproj" PrivateAssets="none" />
```
Do you need to add the feature in the recipes ?

matiasmolleja replied:
Not sure if this should be enabled by default. Just in case you think it's OK, I added a last commit.
It is adding an admin tree to the Blog recipe. Adds a menu for the blog and adds a list of content types menus to the content items menu:

![seed-tree-blog-recipe](https://user-images.githubusercontent.com/2589629/47514665-408a8200-d881-11e8-9c13-45acdb4b2700.jpg)

There is an issue about the blog editing page being difficult to find. This should fix it for the Blog recipe.

sebastienros commented:
Can you add the Blog node as a top level one in the Blog recipe?

matiasmolleja commented:
I liked too at the root level. But can't assign it an icon. At the moment I only can add icons to the link nodes.

sebastienros commented:
But it used to work, you showed it, and you even added an icon picker ... I am confused. Is that a recipe limitation?

matiasmolleja commented:
The icon picker is only on the link admin. Maybe when I showed I still had the "default circle" icon that was added to every root menuitem that had no icon set.

One alternative woud be to add a link with a url that points to the blog edit page. I guess I can build the url by using an blogContentItemId variable that we have on the recipe

sebastienros commented:
Conflicts ,then merge

matiasmolleja commented:
After your review I made some changes:

- Changed look an feel of lists to be more like our menu item
- Changed the way the buttons to enable-disable are shown.
- Added icon picker to every admin node. In the content types admin node, it is possible to pick a different icon for each type.
- Blog Recipe: Added the blog on the root. The same for the main menu content item.
- Added a new admin note type: placeholder . It's like the link type but it has no Url. Only for containing other items.

matiasmolleja commented:
![tree-changes](https://user-images.githubusercontent.com/2589629/48022967-d2757300-e13c-11e8-93f7-462cd646ac23.gif)
